ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -110,9 +131,15 @@ private void initializeProtoArrayForkChoice() {
                       final List<UInt64> justifiedEffectiveBalances =
                           spec.getBeaconStateUtil(justifiedState.getSlot())
                               .getEffectiveBalances(justifiedState);
+
+                      final List<ProposerWeighting> removedProposerWeightings =
+                          proposerWeightings.clearProposerWeightings();
```

Comment:
The timing here should work because we're calling `processHead` at the beginning of the slot?  So any new block for this slot should theoretically come in after that?  Could we make this safer by passing in the current slot and only clearing weightings prior to this slot?

Reply:
Yeah there's two important interactions going on here.  At the end of slot x (ie start of slot x+1) we should:

1. Remove any proposer weightings applied for slot x (or before)
2. Apply any attestations from slot x (and thus not consider any attestations from x received after the end of the slot in choosing the parent for the block in slot x+1)

The timing of applying votes is the key to making both work (we do it at the start of the slot, prior to creating a block).

I had originally thought we'd need to remove only weightings prior to a specific slot but because the current slot is based on current system time we always want to remove them all. If we ever wind up with a proposer weighting from a slot in the future because of a bug we'd want to remove it anyway.

You're right that it is hard to see these interactions fitting together though which is a concern... Can write some docs, but it would be nice to make it clearer in the code, just not sure how.